crategen/validators.py

```diff
@@ -0,0 +1,109 @@
+from pydantic import ValidationError
+from typing import Union
+from .models import WRROCProcess, WRROCWorkflow, WRROCProvenance
+from urllib.parse import urlparse
+
+def validate_wrroc(data: dict) -> Union[WRROCProvenance, WRROCWorkflow, WRROCProcess]:
+    """
+    Validate that the input data is a valid WRROC entity and determine which profile it adheres to.
+    
+    This function attempts to validate the input data against the WRROCProvenance model first.
+    If that validation fails, it attempts validation against the WRROCWorkflow model.
+    If that also fails, it finally attempts validation against the WRROCProcess model.
+    
+    Args:
+        data (dict): The input data to validate.
+    
+    Returns:
+        Union[WRROCProvenance, WRROCWorkflow, WRROCProcess]: The validated WRROC data, indicating the highest profile the data adheres to.
+    
+    Raises:
+        ValueError: If the data does not adhere to any of the WRROC profiles.
+    """
+    # Convert '@id' to 'id' for validation purposes
+    if '@id' in data:
+        data['id'] = data.pop('@id')
+
+    errors = []
+
+    try:
+        return WRROCProvenance(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+
+    try:
+        return WRROCWorkflow(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+
+    try:
+        return WRROCProcess(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+        raise ValueError(f"Invalid WRROC data: {errors}")
```

Comment:
I am not sure the error chaining makes sense to me. Let's say the `@id` field is missing. This would mean that the data is not a WRROC of any type and the validator would, rightly, raise an error. However, the error message would complain about the missing field 3 times.

I think what we should do is just add `pass` for the first two profiles, then, in the last case (`WRROCProcess`), we should simply re-raise the error. This is because we don't know what profile the data is supposed to comply with. So when we find out that the data is not a WRROC at all, we complain about what is missing to make it a _minimal_ valid WRROC, which is `WRROCProcess`.

If we _do_ know the expected profile, then of course we would only check against that, and raise the error that the validation against that particular profile model uncovers. But without that information, I don't think it makes sense to include error messages from the parent models...

So what I would do:

```py
    try:
        return WRROCProvenance(**data)
    except ValidationError as e:
        pass

    try:
        return WRROCWorkflow(**data)
    except ValidationError as e:
        pass

    try:
        return WRROCProcess(**data)
    except ValidationError as exc:
        raise ValueError(f"Invalid WRROC data: {exc.errors()}") from exc
```

You can then of course remove the `errors = []`. By the way, if you define (local) vars, remember to add type hints.

Reply:
I’ve updated the validators to use the correct models (`WRROCDataTES` and` WRROCDataWES)` as specified. This ensures that the validation is aligned with the expected data structures.